File: 백준/Gold/2447. 별 찍기 － 10/별 찍기 － 10.cc
#include <iostream>
using namespace std;
void star(int** arr, int i, int j, int n);
int main() {
	int n;
	cin >> n;
	int** arr = new int* [n];
	for (int i = 0; i < n; i++) {
		arr[i] = new int[n];
		for (int j = 0; j < n; j++) {
			arr[i][j] = 0;
		}
	}
	star(arr, 0, 0, n);
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (arr[i][j] == 1) {
				cout << '*';
			}
			else {
				cout << ' ';
			}
		}
		cout << '\n';
	}
	for (int i = 0; i < n; i++) {
		delete[] arr[i];
	}
	delete[] arr;
	return 0;
}
void star(int** arr, int i, int j, int n) {
	if (n > 3) {
		for (int a = 0; a < 3; a++) {
			for (int b = 0; b < 3; b++) {
				if (!(a == 1 && b == 1)) {
					star(arr, i + a * n / 3, j + b * n / 3, n / 3);
				}
			}
		}
	}
	else {
		for (int a = 0; a < 3; a++) {
			for (int b = 0; b < 3; b++) {
				if (!(a == 1 && b == 1)) {
					arr[i + a][j + b] = 1;
				}
			}
		}
	}
}

File: 백준/Gold/2447. 별 찍기 － 10/별 찍기 － 10.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        char[][] arr = new char[n][n];
        star(arr, n, 0, 0, false);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                sb.append(arr[i][j]);
            }
            sb.append("\n");
        }
        System.out.print(sb);
    }

    public static void star(char[][] arr, int n, int x, int y, boolean blank) {
        if (blank) {
            for (int i = x; i < x + n; i++) {
                for (int j = y; j < y + n; j++) {
                    arr[i][j] = ' ';
                }
            }
            return;
        }
        if (n == 1) {
            arr[x][y] = '*';
            return;
        }
        int size = n / 3;
        for (int i = x; i < x + n; i += size) {
            for (int j = y; j < y + n; j += size) {
                if (i == x + size && j == y + size) {
                    star(arr, size, i, j, true);
                } else {
                    star(arr, size, i, j, false);
                }
            }
        }
    }
}
